src/crimson/os/seastore/object_data_handler.h

```diff
@@ -30,21 +30,63 @@ struct block_delta_t {
   }
 };
 
+struct overwrite_buf_t : public boost::intrusive_ref_counter<
+    overwrite_buf_t,
+    boost::thread_unsafe_counter> {
+  size_t add(block_delta_t &b) {
+    changes.push_back(b);
+    return changes.size();
+  }
+  bool empty() {
+    return changes.empty();
+  }
+  size_t get_size() const {
+    return changes.size();
+  }
+  void clear() {
+    if (changes.size()) {
+      changes.clear();
+    }
+  }
+  void copy_changes(bufferptr &b, int to) {
+    int idx = 0;
+    for (auto p : changes) {
+      auto iter = p.bl.cbegin();
+      iter.copy(p.bl.length(), b.c_str() + p.offset);
+      idx++;
+      if (idx >= to) {
+	break;
+      }
+    }
+  }
+  std::vector<block_delta_t> changes = {};
+};
+using overwrite_buf_ref = boost::intrusive_ptr<overwrite_buf_t>;
+
 struct ObjectDataBlock : crimson::os::seastore::LogicalCachedExtent {
   using Ref = TCachedExtentRef<ObjectDataBlock>;
 
   std::vector<block_delta_t> delta = {};
 
   interval_set<extent_len_t> modified_region;
+  // to provide the local modified view during transaction
+  overwrite_buf_ref overwrites = nullptr; 
+  int overwrites_idx = 0;
```

Comment:
The `overwrites_idx` doesn't seem necessary -- the data extent can only be changed to mutation-pending only once upon the 1st mutation within a transaction, and `overwrite_buf_ref overwrites` is only valid before the transaction is committed.

Or am I missing anything?